torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1242,6 +1245,24 @@ void ProcessGroupNCCL::workCleanupLoop() {
     // Bump up heart beat by one.
     heartbeat_++;
 
+    // poll store to see if some ranks have flagged a timeout when
+    // we haven't polled for `heartbeat_timeout` seconds and there haven't
+    // any work added or removed for `watchdog_timeout` seconds.
+    if (dumpOnTimeout_) {
+      auto currentTimepoint = std::chrono::steady_clock::now();
+      if (std::chrono::duration_cast<std::chrono::milliseconds>(
+              currentTimepoint - lastWorkListUpdateTime_) >=
+              std::chrono::milliseconds(kWatchdogThreadSleepMillis) &&
+          std::chrono::duration_cast<std::chrono::seconds>(
+              currentTimepoint - lastTimePollStore) >=
+              std::chrono::seconds(heartbeatTimeoutInSec_)) {
```

Comment:
i'm unsure why you need to keep track of both lastWorkListUpdateTime_ and lastTimePollStore, and why you use two separate thresholds (kWatchdogThreadSleepMillis and heartbeatTimeoutInSec_).

The 'in english' version of this is really just "has it been a while since we've observed local progress?  if so, lets ping the store and see if anyone else timed out".  Any reason we can't do that with only the `lastWorkListUpdateTime_` portion of the code?

Also, what's your reasoning for how long is long enough?

Reply:
Well, I guess the logic here is to optimize the poll so that we don't frequently read from Store. So for polling we use `heartbeatTimeoutInSec_` so that we wait at least that amount of time to check for the next time. And also if we see work is being added or removed from the list, it's not time yet, either. Well, I think one thing to simplify the logic here is to use only `lastWorkListUpdateTime_` and `heartbeatTimeoutInSec_`. cc: @zdevito to see if he has any comments on my proposal.

I don't have an answer to your second question but it does not matter right? We at least wait for some time (which in this case 2 mins) so that we don't query Store crazily. Let's say in 32K's scenario, query every 2 min, you get around 300 query/sec, which is reasonably low?

Reply:
yea there should be 2 constraints. First we want to not overwhelm the store, Second we want to dump soon enough to avoid being killed before we dump.  But i couldn't say what the values are for either of those constraints.  Maybe 2 min is fine.

Reply:
So what is the conclusion here? I can change the logic and I think this is the only blocker which prevent this PR from merging I guess?

Reply:
my bad. i agree we need both conditions.  (would be nice if we could clean up the way it looks a bit )